bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -184,19 +193,59 @@ public void addRow(ByteString message) throws IOException {
     appendRequestRowCount++;
   }
 
+  /**
+   * Thorws an exception if there is an error in any of the responses received thus far. Since
+   * responses arrive in order, we proceed to check the next response only after the previous
+   * response has arrived.
+   *
+   * @throws BigQueryConnectorException If there was an invalid response
+   */
+  private void checkForFailedResponse(boolean waitForResponse) {
+    while (nextIndexCheckResponse < appendRowsFutures.size()) {
+      ApiFuture<AppendRowsResponse> validatedAppendRowsFuture =
+          appendRowsFutures.get(nextIndexCheckResponse);
+      if (waitForResponse || validatedAppendRowsFuture.isDone()) {
+        boolean succeeded = false;
+        try {
+          AppendRowsResponse response = validatedAppendRowsFuture.get();
+          // If we got here, the response was successful
+          succeeded = true;
+          nextIndexCheckResponse++;
+        } catch (InterruptedException | ExecutionException e) {
+          logger.error(
+              "Write-stream {} with name {} exception while retrieving AppendRowsResponse",
+              partitionId,
+              writeStreamName);
+          throw new BigQueryConnectorException("Exception while retrieving AppendRowsResponse", e);
+        } finally {
+          if (!succeeded) {
+            appendRowsExecutor.shutdown();
+          }
+        }
+      } else {
+        break;
+      }
+    }
+  }
+
   /**
    * Sends an AppendRowsRequest to the BigQuery Storage Write API.
    *
    * @throws IOException If the append rows request fails: either by returning the wrong offset
    *     (deduplication error) or if the response contains an error.
    */
   private void sendAppendRowsRequest() throws IOException {
+    checkForFailedResponse(false);
```

Comment:
```suggestion
    checkForFailedResponse(false /* waitForResponse */);
```
Similarly, for the other call.

Reply:
Done